cEP-0024.md

```diff
@@ -0,0 +1,85 @@
+| Authors  | Vamshi Krishna Bommerla <mailto:bommvams72@gmail.com> |
+| Status   | Proposed                                              |
+| Type     | Process                                               |
+
+## Abstract
+
+This cEP describes the details of the process of creating new automations and
+plugins in GitMate which helps coala development workflow, as a part of the
+[GSoC project](https://summerofcode.withgoogle.com/projects/#5344316227059712).
+
+## Introduction
+
+[GitMate](https://gitmate.io/) is an automation tool for developers. While it
+works well, there area number of plugins which saves a lot of time of our
+community(for example automatically close PRs from master). Many of them are
+general plugins, which are required by every org, but we need more plugins
+which support coala development workflow.
+
+## New Features
+
+### Auto-assign issue based on commit mention
+
+When a commit mentions(fixes, closes) a unassigned issues, auto assign
+commit author to the mentioned unassigned issues.
+
+### Auto-reject PR opened on master of forks
+
+Normally, we should make changes in different branch from master of forks, and
+submit a PR. But, sometimes we receive PRs from master of forks. To solve this
+problem, we need a mechanism by which gitmate automatically rejects those PR’s,
+and inform the PR author to work on a different branch.
+
+### Auto-mark PR with necessary labels as per the reviews done
+
+When a reviewer reviews a PR, he either approves changes or request changes or
+comments something and leaves it there, but PR would be still in the review
+queue, so there should be an automation to mark approve label if reviewer
+approves changes and mark WIP if reviewer requests changes or comments on the
+source code when there is CI failure.
+
+### Fixes vs Closes automated checking
+
+Our members often confuse when we should use `Fixes` or `Closes` on a commit.
+So, GitMate should automatically check every commit for issues which is being
+Fixed or Closed, if those issues have bug label(label will be configurable),
+`Fixes` should be used, and vice versa. If there is an error, gitmate commit
+review should fail and inform the PR author how to fix it.
+
+### Assign Plugin
+
+Assign issue when a user comments on an issue requesting for an `assignment`
+to that issue. Will support settings for each team like limit team to work on
+issues with specific set of labels, limit number of issues of specified label,
+a team member can work on. corobo does all these things, but it is based just
+on gitter messages and corobo is used in coala only, but gitmate is based on
+github, used by many other users and is more configurable.
+
+```
+Settings:
+
+Team - newcomer
+Label - newcomer  Limit - 1
+Label - low       Limit - 2
+
+Team 2....
+```
+
+```
+> @gitmate-bot assign
+
+  @user, you have been assigned
+```
+
+### Auto-review merge conflicts
+
+Check for merge conflicts in a PR, if there are any conflicts, mark PR with
```

Comment:
There are two types of merge conflicts, and it is hard to know which one you are referring to here.

1. Merge conflicts in the commit, which is solved with https://github.com/coala/coala-bears/pull/2244 .

2. Then there is PRs which are *stale* and have merge conflicts with their target branch, e.g. master.

I assume this is about second one.  There are lots of libraries which provide pending merge conflict detection.  GitHub is using one, and they show a lovely message at the bottom of the PR if it cant merged due to pending merge conflicts.

There is also another feature which I think zulip has, which is identifying which PRs will become unmergable if a PR is merged.  i.e. overlapping PRs.  This information can be helpful to ensure that the best PR is merged of the ones which are overlapping, and the others then become broken.

Reply:
@jayvdb We can directly use GitHub to know if the the PR has merge conflicts by `mergeable_state` in pull request.
https://developer.github.com/v3/pulls/#get-a-single-pull-request

Reply:
What about GitLab, then @Vamshi99?

Reply:
Possible GitLab implementation:

Clone the repo into a docker container, and then perform the merge operation, if the merge operation has a clean exit code (0), that means there'd be no conflicts.

Reply:
The text hasnt been updated to clarify this.